doc/design/to_run_stmt.md

```diff
@@ -0,0 +1,178 @@
+
+We can express the data transform logic using Python, leverage many mature python packages and encapsulate it into a reusable function. And then we propose to add the `TO RUN` clause in SQLFlow to call it.
+
+## Syntax Extension
+
+```SQL
+SELECT * FROM source_table
+TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func
+WITH param_a = value_a,
+     param_b = value_b
+INTO result_table
+```
+
+The SQLFlow statement above will call the python function `a_python_func` from the file `/run/my_func/data_preprocessor.py` inside the docker image `a_data_scientist/maxcompute_functions:1.0`. The attributes in the `WITH` clause will be passed to the python function as parameters.
+
+## Challenges
+
+- TO RUN can execute functions in Kubernetes and MaxCompute.
+- TO RUN can execute functions in both a single process and a distributed cluster.
+- TO RUN can execute functions built upon various data computing frameworks such as [Dask](https://github.com/dask/dask), [Mars](https://github.com/mars-project/mars), [Ray](https://github.com/ray-project/ray) and so on.
+
+## Design
+
+### Fix the challenges
+
+The semantics of `TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func` is that `a_python_func` will be executed in a docker container for Kubernetes or a PyODPS node for MaxCompute. The implementation of `a_python_func` is fully customized by users. Because the execution environments for `a_python_func` are different between Kubernetes and MaxCompute, there are some differences considering these three challenges above.
+
+Kubernetes
+
+- Execution environment  
+  `a_python_func` is executed inside the function image `a_data_scientist/maxcompute_functions:1.0`. We can customized this image at the `docker build` stage and make sure that all the dependencies `a_python_func` need are already installed in the image.
+
+- Single process execution  
+  User writes the data transform process using pandas, numpy, tsfresh and so on inside `a_python_func`.
+
+- Distributed cluster execution  
+  User can install Dask/Mars/Ray while building the docker image. Inside the implementation of `a_python_func`, we can create a Dask/Mars/Ray cluster, build a computing graph and submit the distribution execution using the API from these packages.
+
+MaxCompute
```

Comment:
I don't think there is a configuration of PyODPS node on MaxCompute. In my mind, there is only one execution environment -- Docker container on Kubernetes.  It is the function's responsibility to launch a job on MaxCompute.

Reply:
Yes, we can only execute docker run in the workflow step.
For different platforms, the commands are different for the docker run: `python a_python_file.py` for Vanilla Kubernetes and `alisa.submitter a_python_file.py` for MaxCompute.
It's the responsibility of `alisa.submitter` to submit a PyODPS task to MaxCompute to execute the Python program.